### PR TITLE
[Networking] Tuning libp2p resource manager limits; supporting by tests

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR implements a solution to mitigate the libp2p goroutine leakage in production. 
- Rigorous testing implemented for inbound stream resource management that was causing the goroutine leakage. 
- Two principal assumptions have been examined: (1) we can limit the number of streams per peer (2) we can limit the number of streams at the system-wide scale. 
- Hard configurable limits have been put on the inbound streams notably at the system and peer level to suppress inbound stream overload cases. 
- Edge cases have been examined for the resource manager configuration and failing edge cases have been re-produced as (quarantined) tests to be investigated further and shared with the libp2p team.

### CODE REVIEW

File: network/internal/p2putils/utils.go [hunk 1]
```diff
@@ -69,46 +69,95 @@ func ConnectednessToString(connectedness network.Connectedness) (string, bool) {
...
+//     stream := FilterStream(host, targetID, Direction(network.DirInbound))
+//
+// host is the host from which to filter streams.
+// targetID is the ID of the target peer.
+// options is a variadic parameter that allows zero or more FilterOption values to be provided.
+//
+// It returns a slice of network.Stream values that match the filtering criteria.
+func FilterStream(host host.Host, targetID peer.ID, options ...FilterOption) []network.Stream {
 	var filteredStreams []network.Stream
+	const discardTheProtocol = "discard-the-protocol"
```
Comment (kc1116) [inline network/internal/p2putils/utils.go:136]:
```suggestion
	const allProtocols = "*"
```

Comment (kc1116) [inline network/internal/p2putils/utils.go:136]:
```suggestion
	const allProtocols = "*"
```

File: network/p2p/p2pnode/resourceManager_test.go [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+func TestCreateStream_MaxProtocolLimit(t *testing.T) {
+	base := baseCreateStreamInboundStreamResourceLimitConfig()
+	base.maxInboundStreamProtocol = math.MaxInt
+	testCreateStreamInboundStreamResourceLimits(t, base)
+}
+
+func TestCreateStream_MinPeerProtocolLimit(t *testing.T) {
+	unittest.SkipUnless(t,
+		unittest.TEST_TODO,
+		"max inbound stream peer protocol is not preserved; can be partially due to count steam not counting inbound streams on a protocol")
```
Comment (kc1116) [inline network/p2p/p2pnode/resourceManager_test.go:180]:
```suggestion
		"max inbound stream peer protocol is not preserved; can be partially due to count stream not counting inbound streams on a protocol")
```

File: network/p2p/p2pnode/resourceManager_test.go [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+func TestCreateStream_MaxPeerLimit(t *testing.T) {
+	base := baseCreateStreamInboundStreamResourceLimitConfig()
+	base.maxInboundPeerStream = math.MaxInt
+	testCreateStreamInboundStreamResourceLimits(t, base)
+}
+
+func TestCreateStream_MinProtocolLimit(t *testing.T) {
+	unittest.SkipUnless(t,
+		unittest.TEST_TODO,
+		"max inbound stream protocol is not preserved; can be partially due to count steam not counting inbound streams on a protocol")
```
Comment (kc1116) [inline network/p2p/p2pnode/resourceManager_test.go:165]:
```suggestion
		"max inbound stream protocol is not preserved; can be partially due to count stream not counting inbound streams on a protocol")
```

File: network/p2p/p2pnode/resourceManager_test.go [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+	base.maxInboundPeerStream = 5          // each peer can only create 5 streams.
+	base.maxInboundStreamPeerProtocol = 10 // each peer can create 10 streams on a specific protocol (but should still be limited by the peer-level limit).
+	testCreateStreamInboundStreamResourceLimits(t, base)
+}
+
+func TestCreateStream_PeerLimitGreaterThanPeerProtocolLimit(t *testing.T) {
+	// the case where peer-level limit is higher than the peer-protocol-level limit.
+	unittest.SkipUnless(t,
+		unittest.TEST_TODO,
+		"max inbound stream peer protocol is not preserved; can be partially due to count steam not counting inbound streams on a protocol")
```
Comment (kc1116) [inline network/p2p/p2pnode/resourceManager_test.go:240]:
```suggestion
		"max inbound stream peer protocol is not preserved; can be partially due to count stream not counting inbound streams on a protocol")
```

File: network/p2p/p2pnode/resourceManager_test.go [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+	base.maxInboundStreamProtocol = 100   // overall limit is 100 streams on a specific protocol (across all peers).
+	base.maxInboundStreamTransient = 1000 // overall limit is 1000 transient streams.
+	base.maxInboundStreamSystem = 1000    // overall limit is 1000 system-wide streams.
+	testCreateStreamInboundStreamResourceLimits(t, base)
+}
+
+func TestCreateStream_ProtocolLimitLessThanPeerProtocolLimit(t *testing.T) {
+	unittest.SkipUnless(t,
+		unittest.TEST_TODO,
+		"max inbound stream peer protocol is not preserved; can be partially due to count steam not counting inbound streams on a protocol")
```
Comment (kc1116) [inline network/p2p/p2pnode/resourceManager_test.go:253]:
```suggestion
		"max inbound stream peer protocol is not preserved; can be partially due to count stream not counting inbound streams on a protocol")
```

**Review by kc1116 — APPROVED** [review]

File: config/default-config.yml [hunk 6]
```diff
@@ -58,6 +58,19 @@ network-config:
...
+  # maximum number of inbound transient limit for streams, across all streams that are not yet fully opened and associated with a protocol.
+  # Note that streams are ephemeral and are created and destroyed intermittently.
+  libp2p-inbound-stream-limit-transient: 15_000
+  # maximum number of inbound limit for streams for each protocol across all peers; this is a per-protocol limit. We expect at least
+  # three protocols per node; gossipsub, unicast, and dht. Note that streams are ephemeral and are created and destroyed intermittently.
+  libp2p-inbound-stream-limit-protocol: 5000
+  # maximum number of inbound streams from each peer across all protocols.
+  libp2p-inbound-stream-limit-peer: 1000
+  # maximum number of inbound streams from each peer for each protocol.
+  libp2p-inbound-stream-limit-protocol-peer: 500
```
Comment (peterargue) [inline config/default-config.yml:73]:
```suggestion
  # maximum number of inbound system-wide streams, across all peers and protocols
  # Note that streams are ephemeral and are created and destroyed intermittently.
  libp2p-inbound-stream-limit-system: 15_000
  # maximum number of inbound transient streams, across all streams that are not yet fully opened and associated with a protocol.
  # Note that streams are ephemeral and are created and destroyed intermittently.
  libp2p-inbound-stream-limit-transient: 15_000
  # maximum number of inbound streams for each protocol across all peers; this is a per-protocol limit. We expect at least
  # three protocols per node; gossipsub, unicast, and dht. Note that streams are ephemeral and are created and destroyed intermittently.
  libp2p-inbound-stream-limit-protocol: 5000
  # maximum number of inbound streams from each peer across all protocols.
  libp2p-inbound-stream-limit-peer: 1000
  # maximum number of inbound streams from each peer for each protocol.
  libp2p-inbound-stream-limit-protocol-peer: 500
```

**Review by peterargue — APPROVED** [review]

File: config/default-config.yml [hunk 7]
```diff
@@ -58,6 +58,19 @@ network-config:
   #	Without this limit peers can end up in a state where there exists n number of connections per peer which
   #	can lead to resource exhaustion of the libp2p node.
   libp2p-peer-base-limits-conns-inbound: 1
+  # maximum number of inbound system-wide streams, across all peers and protocols
+  # Note that streams are ephemeral and are created and destroyed intermittently.
+  libp2p-inbound-stream-limit-system: 15_000
+  # maximum number of inbound transient streams, across all streams that are not yet fully opened and associated with a protocol.
+  # Note that streams are ephemeral and are created and destroyed intermittently.
+  libp2p-inbound-stream-limit-transient: 15_000
```
Comment (gomisha) [inline config/default-config.yml:66]:
```suggestion
  libp2p-inbound-stream-limit-transient: 15000
```

File: config/default-config.yml [hunk 8]
```diff
@@ -58,6 +58,19 @@ network-config:
   #	Without this limit peers can end up in a state where there exists n number of connections per peer which
   #	can lead to resource exhaustion of the libp2p node.
   libp2p-peer-base-limits-conns-inbound: 1
+  # maximum number of inbound system-wide streams, across all peers and protocols
+  # Note that streams are ephemeral and are created and destroyed intermittently.
+  libp2p-inbound-stream-limit-system: 15_000
```
Comment (gomisha) [inline config/default-config.yml:63]:
```suggestion
  libp2p-inbound-stream-limit-system: 15000
```

File: network/p2p/p2pbuilder/libp2pscaler.go [hunk 9]
```diff
@@ -16,14 +16,14 @@ func getNumFDs() (int, error) {
 	return int(l.Cur), nil
 }
 
-func allowedMemory(scaleFactor float64) (int64, error) {
+func AllowedMemory(scaleFactor float64) (int64, error) {
```
Comment (gomisha) [inline network/p2p/p2pbuilder/libp2pscaler.go:19]:
```suggestion
func allowedMemory(scaleFactor float64) (int64, error) {
```

Tests can reference and run the unexported version of this function so why the need to export it?

File: network/p2p/p2pbuilder/libp2pscaler.go [hunk 10]
```diff
@@ -16,14 +16,14 @@ func getNumFDs() (int, error) {
...
 	if scaleFactor <= 0 || scaleFactor > 1 {
 		return 0, fmt.Errorf("memory scale factor must be greater than 0 and less than or equal to 1: %f", scaleFactor)
 	}
 	return int64(math.Floor(float64(memory.TotalMemory()) * scaleFactor)), nil
 }
 
-func allowedFileDescriptors(scaleFactor float64) (int, error) {
+func AllowedFileDescriptors(scaleFactor float64) (int, error) {
```
Comment (gomisha) [inline network/p2p/p2pbuilder/libp2pscaler.go:26]:
```suggestion
func allowedFileDescriptors(scaleFactor float64) (int, error) {
```
Tests can reference and run the unexported version of this function so why the need to export it?

File: network/p2p/p2pbuilder/utils.go [hunk 11]
```diff
@@ -32,8 +33,8 @@ type limitConfigLogger struct {
 	logger zerolog.Logger
 }
 
-// newLimitConfigLogger creates a new limitConfigLogger.
-func newLimitConfigLogger(logger zerolog.Logger) *limitConfigLogger {
+// NewLimitConfigLogger creates a new limitConfigLogger.
+func NewLimitConfigLogger(logger zerolog.Logger) *limitConfigLogger {
```
Comment (gomisha) [inline network/p2p/p2pbuilder/utils.go:37]:
```suggestion
// newLimitConfigLogger creates a new limitConfigLogger.
func newLimitConfigLogger(logger zerolog.Logger) *limitConfigLogger {
```

Why export this outside the package? It's not used outside the package.

File: config/default-config.yml
Comment (gomisha) [inline config/default-config.yml:1]: We should try to be consistent and make all values > 999 use underscore:

e..g `libp2p-inbound-stream-limit-protocol: 5000` => `libp2p-inbound-stream-limit-peer: 5_000`
e.g. `gossipsub-rpc-inspector-notification-cache-size: 10000` => `gossipsub-rpc-inspector-notification-cache-size: 10_000`

**Review by gomisha — APPROVED** [review]
Looks good, just a few minor nits and suggestions.

File: network/p2p/p2pbuilder/libp2pscaler.go (hunk 10 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/libp2pscaler.go]: https://github.com/onflow/flow-go/pull/4846/commits/3bbaf2470a3825ea4e1ee72f7bc468c4870b3867

File: network/p2p/p2pbuilder/libp2pscaler.go (hunk 9 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/libp2pscaler.go]: https://github.com/onflow/flow-go/pull/4846/commits/3bbaf2470a3825ea4e1ee72f7bc468c4870b3867

File: network/p2p/p2pbuilder/utils.go (hunk 11 above)
Comment (yhassanzadeh13) [reply network/p2p/p2pbuilder/utils.go]: https://github.com/onflow/flow-go/pull/4846/commits/0de5c5d81e0fe4f7566e07cd4f383b5d0ec70966

File: config/default-config.yml
Comment (yhassanzadeh13) [reply config/default-config.yml:1]: https://github.com/onflow/flow-go/pull/4846/commits/dd83fe6a55a89ee813a23484e8f7d528c3e56354

**Pull request merged**